setuptools/command/bdist_wheel.py

```diff
@@ -433,7 +442,15 @@ def run(self):
             f"{safer_version(self.distribution.get_version())}.dist-info"
         )
         distinfo_dir = os.path.join(self.bdist_dir, distinfo_dirname)
-        self.egg2dist(self.egginfo_dir, distinfo_dir)
+        if self.dist_info_dir:
+            # Use the given dist-info directly.
+            shutil.copytree(self.dist_info_dir, distinfo_dir)
+            # Egg info is still generated, so remove it now to avoid it getting
+            # copied into the wheel.
```

Comment:
Do we know why egg info is still generated? Is it because other commands might have generated it? What happens if the line for `rmtree` is removed (will the tests fail)?

While I'm okay with this approach, I'm also somewhat apprehensive, as it is perpetuating and reinforcing the expectation that egg-info is generated, even when it's not needed, making it harder in the future to correct this suboptimal approach. For example, this project also has the goal to someday obviate egg-info altogether, and to generate dist-info directly, and adding this additional handling of egg-info will likely complicate that work (though maybe not too much).

I wonder - how much do other commands from `bdist_wheel` depend on the generation of egg-info and could those commands get by without generating it if `bdist_wheel` could pass through the instruction to re-use existing dist-info?

Reply:
> Do we know why egg info is still generated?

Looks like `install` calls `install_egg_info` (a registered sub-command, which is automatically run), which then calls `egg_info`. I wasn't brave enough to try to touch this.

> What happens if the line for rmtree is removed (will the tests fail)?

I've added a specific test for this. Previously the test would have passed.

> this project also has the goal to someday obviate egg-info altogether

Yes, I appreciate that. IMO, this change does add an additional place where you have to think about egg-info, but I believe it is already necessary to deal with `bdist_wheel` wrt `egg-info`, and the added code-path now explicitly rejects its use - I think that ultimately makes that easier to remove egg-info in the future.